File: golang/0199-binary-tree-right-side-view.go
package main

import (
	"fmt"
)

func main() {
	n := TreeNode{Val: 1, Left: &TreeNode{Val: 2, Left: nil, Right: &TreeNode{Val: 5, Left: nil, Right: nil}}, Right: &TreeNode{Val: 3, Left: nil, Right: &TreeNode{Val: 4, Left: nil, Right: nil}}}
	fmt.Println(rightSideView(&n))
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rightSideView(root *TreeNode) []int {
	result := make([]int, 0)
	rightView(root, &result, 0)
	return result
}

func rightView(curr *TreeNode, result *[]int, currDepth int) {
	if curr == nil {
		return
	}
	if currDepth == len(*result) {
		*result = append(*result, curr.Val)
	}

	rightView(curr.Right, result, currDepth+1)
	rightView(curr.Left, result, currDepth+1)
}

File: python/0199-Binary-Tree-Right-Side-View.py
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        result = []
        self.dfs(root, result, 0)
        return result

    def dfs(self, node: TreeNode, result: List[int], depth: int) -> None:
        if depth == len(result):
            result.append(node.val)
        for ch in [node.right, node.left]:
            if ch:
                self.dfs(ch, result, depth + 1)


if __name__ == "__main__":
    s = Solution()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(7)
    root.right.left.right = TreeNode(8)

    print(s.rightSideView(root))
